File: python/robot-simulator/robot_simulator.py
# Globals for the directions
# Change the values as you see fit
EAST = 'east'
NORTH = 'north'
WEST = 'west'
SOUTH = 'south'

dlist = [NORTH,EAST,SOUTH,WEST]

class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, directions):
        for d in [*directions]:
            if d == 'R':
                self.direction = dlist[(dlist.index(self.direction) + 1) % 4]
            elif d == 'L':
                self.direction = dlist[(dlist.index(self.direction) - 1)]
            elif d == 'A':
                cur = self.coordinates
                match self.direction:
                    case 'north':
                        self.coordinates = (cur[0], cur[1] + 1)
                    case 'east':
                        self.coordinates = (cur[0] + 1, cur[1])
                    case 'south':
                        self.coordinates = (cur[0], cur[1] - 1)
                    case 'west':
                        self.coordinates = (cur[0] - 1, cur[1])
                    case _:
                        raise ValueError("invalid direction")

File: ruby/robot-simulator/robot_simulator.rb
class Simulator
  MAP = {A: :advance, L: :turn_left, R: :turn_right}.freeze

  def place(robot, x:, y:, direction:)
    robot.at(x,y)
    robot.orient(direction)
  end

  def instructions(input); input.chars.map { |c| MAP[c.to_sym] } end
  def evaluate(robot, input); instructions(input).each { |i| robot.send(i) } end

end

class Robot
  DIRS = %i[north east south west].freeze
  MOVES = DIRS.zip([ [0,1], [1,0], [0,-1], [-1,0] ]).to_h.freeze

  attr_reader :coordinates, :bearing

  def at(x, y); @coordinates = [x, y] end
  def advance; (0..1).each { |i| @coordinates[i] += MOVES[@bearing][i] } end

  def orient(dir)
    raise ArgumentError unless DIRS.include? dir
    @bearing = dir
  end

  def turn(dir); @bearing = DIRS[(DIRS.index(@bearing) + dir) % 4] end
  def turn_right; turn(1) end
  def turn_left; turn(-1) end
end
